File: crates/acpus-core/src/schedule.rs
use crate::{AcpusIr, IrNode, ScheduleBranch, ScheduleNode, ScheduleSummary};

pub fn create_schedule(ir: &AcpusIr) -> ScheduleSummary {
    ScheduleSummary {
        workflow: ir.name.clone(),
        nodes: ir.root.children.iter().map(schedule_node).collect(),
    }
}

fn schedule_node(node: &IrNode) -> ScheduleNode {
    ScheduleNode {
        id: node.id.clone(),
        kind: node.kind.clone(),
        node_path: node.node_path.join("/"),
        output_merge: node.output_merge.clone(),
        children: node.children.iter().map(schedule_node).collect(),
        branches: node
            .branches
            .iter()
            .map(|branch| ScheduleBranch {
                id: branch.id.clone(),
                when: branch.when.clone(),
                children: vec![schedule_node(&branch.child)],
            })
            .collect(),
    }
}

File: legacy/packages/core/src/schedule.ts
import type { AcpusIr, IrNode, ScheduleNode, ScheduleSummary } from "./types.js";

export function createSchedule(ir: AcpusIr): ScheduleSummary {
  return {
    workflow: ir.name,
    nodes: ir.root.children?.map(toScheduleNode) ?? []
  };
}

function toScheduleNode(node: IrNode): ScheduleNode {
  return {
    id: node.id,
    kind: node.kind,
    nodePath: node.nodePath.join("/"),
    outputMerge: node.outputMerge,
    children: node.children?.map(toScheduleNode),
    branches: node.branches?.map((branch) => ({
      id: branch.id,
      when: branch.when,
      children: [toScheduleNode(branch.child)]
    }))
  };
}
